### PR TITLE
[SPARK-41961][SQL] Support table-valued functions with LATERAL

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?
This PR allows table-valued functions to reference columns and aliases in the preceding FROM items using the LATERAL keyword.

### Why are the changes needed?
To improve the usability of table-valued functions.

### Does this PR introduce _any_ user-facing change?
Yes. Before this PR, users cannot invoke a table-valued function in the FROM clause with LATERAL:
```
SELECT * FROM t, LATERAL EXPLODE(ARRAY(t.c1, t.c2));
[INVALID_SQL_SYNTAX] Invalid SQL syntax: LATERAL can only be used with subquery.
```
After this PR, this query can run successfully.

### How was this patch tested?

New SQL query tests.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out [hunk 1]
```diff
@@ -837,6 +837,132 @@ struct<c1:int,c2:array<int>,c3:int>
...
+SELECT * FROM t1, LATERAL (SELECT t1.c1 + c3 FROM EXPLODE(ARRAY(c1, c2)) t(c3) WHERE t1.c2 > 1)
+-- !query schema
+struct<c1:int,c2:int,(outer(spark_catalog.default.t1.c1) + c3):int>
+-- !query output
+1	2	2
+1	2	3
+
+
+-- !query
+SELECT * FROM t1, LATERAL (SELECT * FROM EXPLODE(ARRAY(c1, c2)) l(x) JOIN EXPLODE(ARRAY(c2, c1)) r(y) ON x = y)
```
Comment (sigmod) [inline sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out:956]: Does it make sense to add a SQL query into SchemaPruningSuite for "... from lateral explode .. " or "... from explode ..." to make sure schema pruning behavior is the same: https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaPruningSuite.scala

Or alternatively, add a plan unit test to assert  "... from Lateral explode(...)" and "... from explode(...)" is always compiled to a plan with `Generate`?

Comment (allisonwang-db) [reply sql/core/src/test/resources/sql-tests/results/join-lateral.sql.out:956]: Make sense! I will add more query plan tests.

**Review by sigmod — APPROVED** [review]
thanks @allisonwang-db !

**Comment by allisonwang-db** [comment]
cc @cloud-fan

File: sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql [hunk 2]
```diff
@@ -177,6 +177,25 @@ SELECT * FROM t3 JOIN LATERAL (SELECT EXPLODE_OUTER(c2));
 SELECT * FROM t3 JOIN LATERAL (SELECT EXPLODE(c2)) t(c3) ON c1 = c3;
 SELECT * FROM t3 LEFT JOIN LATERAL (SELECT EXPLODE(c2)) t(c3) ON c1 = c3;
 
+-- SPARK-41961: lateral join with table-valued functions
+SELECT * FROM LATERAL EXPLODE(ARRAY(1, 2));
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql:181]: interesting, do other databases also support this syntax?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeOneRowRelationSubquerySuite.scala [hunk 3]
```diff
@@ -177,4 +177,27 @@ class OptimizeOneRowRelationSubquerySuite extends PlanTest {
...
+    comparePlans(
+      Optimize.execute(query2.analyze),
+      t3.generate(Explode($"arr")).select($"a", $"b", $"arr", $"col".as("v")).analyze)
+
+    // SELECT col FROM t3 JOIN LATERAL (SELECT * FROM EXPLODE(arr) WHERE col > 0)
+    val query3 = t3.lateralJoin(
+      UnresolvedTableValuedFunction("explode", $"arr" :: Nil).where($"col" > 0))
+    val optimized = Optimize.execute(query3.analyze)
+    optimized.exists(_.isInstanceOf[Generate])
+    assertHasDomainJoin(optimized)
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/OptimizeOneRowRelationSubquerySuite.scala:201]: Does this mean we can't run this query today as we can't decorrelate Generate?

File: sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql (hunk 2 above)
Comment (allisonwang-db) [reply sql/core/src/test/resources/sql-tests/inputs/join-lateral.sql:181]: Yes for example Postgres can run this query:
```
select * from lateral unnest(array[1, 2, 3]);
```

**Review by cloud-fan — APPROVED** [review]
LGTM if tests pass

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**